NetworkPkg/Ip6Dxe/Ip6Impl.c

```diff
@@ -63,10 +63,17 @@ EfiIp6GetModeData (
 
   OldTpl     = gBS->RaiseTPL (TPL_CALLBACK);
   IpInstance = IP6_INSTANCE_FROM_PROTOCOL (This);
-  IpSb       = IpInstance->Service;
-  IpIf       = IpInstance->Interface;
+
+  if (IpInstance == NULL) {
+    gBS->RestoreTPL (OldTpl);
+    return EFI_INVALID_PARAMETER;
+  }
+
+  IpSb = IpInstance->Service;
+  IpIf = IpInstance->Interface;
 

```

Comment:
This looks incomplete.
The check for This being NULL is already performed.

If This is not NULL, then it is a very odd corner case for IpInstance to be NULL because IpInstance uses a CR() macro to compute its pointer value from This.  So This must be somewhere in 0 page for this check for NULL to be true.
```
#define IP6_INSTANCE_FROM_PROTOCOL(Ip6) \
          CR ((Ip6), IP6_PROTOCOL, Ip6Proto, IP6_PROTOCOL_SIGNATURE)
```

In addition, if This and IpInstance are not NULL, then IpSb and IpIf are read from IpInstance fields and are immediately used as pointer without checking for those fields being NULL.  Why check for this extremely rare corner case for IpInstance NULL and not check the likely higher probability of a field of IpInstance being NULL???

Reply:
Yeah, I agree this is a very rare corner case. The real fix in this file is line 76, I did a blanket fix across NetworkPkg for returning without restoring the TPL first, as this was a bug reported to me. I had just done the change to the CR macro, so I added that here, also. I'm fine to drop this check, let me know. Otherwise I'll add checks for IpSb and IpIf.